### PR TITLE
8266631: StandardJavaFileManager: getJavaFileObjects() impl violates the spec

### PR DESCRIPTION
**Pull request by lahodaj**
[a copy of https://github.com/openjdk/jdk/pull/4360 for openjdk/jdk17]

This improves javadoc/specification of several StandardJavaFileManager methods, as requested in these bugs:
https://bugs.openjdk.java.net/browse/JDK-8266631
https://bugs.openjdk.java.net/browse/JDK-8266596
https://bugs.openjdk.java.net/browse/JDK-8266591
https://bugs.openjdk.java.net/browse/JDK-8266590

The CSR is here:
https://bugs.openjdk.java.net/browse/JDK-8268260

/issue JDK-8266631
/issue JDK-8266596
/issue JDK-8266591
/issue JDK-8266590

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issues
 * [JDK-8266631](https://bugs.openjdk.java.net/browse/JDK-8266631): StandardJavaFileManager: getJavaFileObjects() impl violates the spec
 * [JDK-8266596](https://bugs.openjdk.java.net/browse/JDK-8266596): StandardJavaFileManager: default impls of setLocationFromPaths(), getJavaFileObjectsFromPaths() methods don't throw IllegalArgumentException as specified
 * [JDK-8266591](https://bugs.openjdk.java.net/browse/JDK-8266591): StandardJavaFileManager::getJavaFileObjectsFromPaths() methods contain a typo in their spec
 * [JDK-8266590](https://bugs.openjdk.java.net/browse/JDK-8266590): StandardJavaFileManager::setLocationFromPaths() spec contains an error

### Reviewers
 * [Vicente Romero](https://openjdk.java.net/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to 3136b328d536cbaa66a199009fc0542275139d8c
 * [Jonathan Gibbons](https://openjdk.java.net/census#jjg) (@jonathan-gibbons - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk17 pull/56/head:pull/56` \
`$ git checkout pull/56`

Update a local copy of the PR: \
`$ git checkout pull/56` \
`$ git pull https://git.openjdk.java.net/jdk17 pull/56/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 56`

View PR using the GUI difftool: \
`$ git pr show -t 56`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk17/pull/56.diff">https://git.openjdk.java.net/jdk17/pull/56.diff</a>

</details>

### CODE REVIEW

**Review by vicente-romero-oracle — APPROVED** [review]
lgtm

File: src/java.compiler/share/classes/javax/tools/StandardJavaFileManager.java [hunk 1]
```diff
@@ -190,16 +190,16 @@
      * Returns file objects representing the given paths.
      *
      * @implSpec
-     * The default implementation converts each path to a file and calls
-     * {@link #getJavaFileObjectsFromFiles getJavaObjectsFromFiles}.
-     * IllegalArgumentException will be thrown if any of the paths
-     * cannot be converted to a file.
+     * The default implementation lazily converts each path to a file and calls
+     * {@link #getJavaFileObjectsFromFiles(Iterable) getJavaFileObjectsFromFiles}.
+     * {@linkplain IllegalArgumentException IllegalArgumentException} will be thrown
```
Comment (jonathan-gibbons) [inline src/java.compiler/share/classes/javax/tools/StandardJavaFileManager.java:195]: Suggest using with `{@link}` or `{@code}` but not `{@linkplain}`.
Given that there is a `@throws` which will link it, `{@code}` may be enough.

File: src/java.compiler/share/classes/javax/tools/StandardJavaFileManager.java [hunk 2]
```diff
@@ -212,10 +212,10 @@
      * Returns file objects representing the given paths.
      *
      * @implSpec
-     * The default implementation converts each path to a file and calls
-     * {@link #getJavaFileObjectsFromFiles getJavaObjectsFromFiles}.
-     * IllegalArgumentException will be thrown if any of the paths
-     * cannot be converted to a file.
+     * The default implementation lazily converts each path to a file and calls
+     * {@link #getJavaFileObjectsFromPaths(Collection) getJavaFileObjectsFromPaths}.
+     * {@linkplain IllegalArgumentException IllegalArgumentException} will be thrown
```
Comment (jonathan-gibbons) [inline src/java.compiler/share/classes/javax/tools/StandardJavaFileManager.java:217]: ditto use `{@code}`

File: src/java.compiler/share/classes/javax/tools/StandardJavaFileManager.java [hunk 3]
```diff
@@ -262,10 +263,15 @@
      *     getJavaFileObjectsFromPaths({@linkplain java.util.Arrays#asList Arrays.asList}(paths))
      * </pre>
      *
+     * @implSpec
+     * The default implementation will only throw {@linkplain NullPointerException NullPointerException}
```
Comment (jonathan-gibbons) [inline src/java.compiler/share/classes/javax/tools/StandardJavaFileManager.java:267]: `{@linkplain}` should only be used for natural-language text. 
Alternatively, it *may* work to use `{@linkplain NullPointerException {@code NullPointerException}}` but that seems overkill.

File: src/java.compiler/share/classes/javax/tools/StandardJavaFileManager.java [hunk 4]
```diff
@@ -332,10 +338,11 @@ void setLocation(Location location, Iterable<? extends File> files)
      * will be cancelled.
      *
      * @implSpec
-     * The default implementation converts each path to a file and calls
-     * {@link #getJavaFileObjectsFromFiles getJavaObjectsFromFiles}.
+     * The default implementation lazily converts each path to a file and calls
+     * {@link #setLocation setLocation}.
      * {@linkplain IllegalArgumentException IllegalArgumentException}
```
Comment (jonathan-gibbons) [inline src/java.compiler/share/classes/javax/tools/StandardJavaFileManager.java:343]: Ditto.

**Review by jonathan-gibbons — APPROVED** [review]
OK ... but please avoid using `{@linkplain}` for code constructs.  Use `{@link}` or `{@code}`

**Review by jonathan-gibbons — APPROVED** [review]

**Pull request closed without merging**